c/tskit/trees.c

```diff
@@ -2076,19 +2076,651 @@ typedef struct {
+            (*num_alleles)++;
+        }
+
+        /* Add the mutation's samples to this allele */
+        tsk_bit_array_get_row(allele_samples, allele, &allele_samples_row);
+        tsk_bit_array_get_row(state, mutation_index, &state_row);
+        tsk_bit_array_add(&allele_samples_row, &state_row);
+
+        /* Get the index for the alternate allele that we must substract from */
+        alt_allele = site->ancestral_state;
+        alt_allele_length = site->ancestral_state_length;
+        if (mutation.parent != TSK_NULL) {
+            parent_mut = site->mutations[mutation.parent - site->mutations[0].id];
+            alt_allele = parent_mut.derived_state;
+            alt_allele_length = parent_mut.derived_state_length;
+        }
+        allele = 0;
+        while (allele < *num_alleles) {
+            if (alt_allele_length == allele_lengths[allele]
+                && tsk_memcmp(alt_allele, alleles[allele], allele_lengths[allele])
+                       == 0) {
+                break;
+            }
+            allele++;
+        }
+        tsk_bug_assert(allele < *num_alleles);
+
+        tsk_bit_array_get_row(allele_samples, allele, &alt_allele_samples_row);
+        tsk_bit_array_subtract(&alt_allele_samples_row, &allele_samples_row);
+    }
+out:
+    tsk_safe_free(alleles);
+    tsk_safe_free(allele_lengths);
+    return ret;
+}
+
+static int
+get_mutation_samples(const tsk_treeseq_t *self, const tsk_size_t tree_index,
+    const tsk_id_t *right_child, const tsk_id_t *left_sib, const tsk_id_t *parent,
+    const tsk_size_t *site_offsets, tsk_bit_array_t *allele_samples,
+    tsk_size_t *restrict *num_alleles)
+// TODO: the restricts here, not sure. Are the row declarations aliases? I think so.
+{
+    int ret = 0;
+
+    const tsk_size_t num_edges = tsk_treeseq_get_num_edges(self);
+    const tsk_size_t num_samples = tsk_treeseq_get_num_samples(self);
+    const tsk_size_t num_sites = self->tree_sites_length[tree_index];
+    const tsk_flags_t *restrict flags = self->tables->nodes.flags;
+    const tsk_site_t *restrict tree_sites = self->tree_sites[tree_index];
+    const tsk_id_t *restrict mut_nodes = self->tables->mutations.node;
+    const tsk_site_t *restrict site;
+
+    tsk_id_t u, node, top_mut_node, *stack;
+    tsk_bit_array_t mut_samples, mut_samples_row, node_paths, path, out_row;
+    tsk_size_t num_mutations, mut_samples_offset, mut_offset;
+    int stack_top;
+
+    // Get the number of mutations and the top mutation node
+    num_mutations = 0;
+    top_mut_node = 0;
+    for (tsk_size_t s = 0; s < num_sites; s++) {
+        site = &tree_sites[s];
+        for (tsk_size_t m = 0; m < site->mutations_length; m++) {
+            num_mutations++;
+            if (site->mutations[m].node > top_mut_node) {
+                top_mut_node = site->mutations[m].node;
+            }
+        }
+    }
+
+    tsk_bit_array_init(&mut_samples, num_samples, num_mutations);
```

Comment:
I see, freeing an uninitialized pointer is undefined behavior. I've added a memset to the initialization function. However, it does seem as though `tsk_safe_free` handles this by setting the pointer to `NULL` first. Since we're initializing these arrays, I have used `free` in the `tsk_bit_array_free` function instead of `tsk_safe_free`.

Reply:
See point in just-made comment about additional safety concerns